rust/lance-linalg/src/distance/cosine.rs

```diff
@@ -348,13 +420,43 @@ mod tests {
         assert_relative_eq!(d[0], 1.0 - 0.873_580_63);
     }
 
+    #[test]
+    fn test_cosine16() {
+        let x: Float16Array = (1..9).map(|v| f16::from_i32(v).unwrap()).collect();
+        let y: Float16Array = (100..108).map(|v| f16::from_i32(v).unwrap()).collect();
+        let d = cosine_distance_batch(x.values(), y.values(), x.len()).collect::<Vec<_>>();
+        assert_relative_eq!(d[0], 1.0);
+
+        let x: Float16Array = [1, 1, 0]
+            .iter()
+            .map(|v| f16::from_i32(*v).unwrap())
+            .collect();
+        let y: Float16Array = [0, 1, 0]
+            .iter()
+            .map(|v| f16::from_i32(*v).unwrap())
+            .collect();
+        let d = cosine_distance_batch(x.values(), y.values(), x.len()).collect::<Vec<_>>();
+        assert_relative_eq!(d[0], 1.0 - 0.707_182_3, epsilon = f32::from(f16::EPSILON));
+
+        let x_val = [3.0, 45.0, 7.0, 2.0, 5.0, 20.0, 13.0, 12.0];
+        let y_val = [2.0, 54.0, 13.0, 15.0, 22.0, 34.0, 50.0, 1.0];
+        let x: Float16Array = x_val.iter().map(|v| f16::from_f32(*v)).collect();
+        let y: Float16Array = y_val.iter().map(|v| f16::from_f32(*v)).collect();
+        let d = cosine_distance_batch(x.values(), y.values(), x.len()).collect::<Vec<_>>();
+        assert_relative_eq!(d[0], 1.0 - 0.873_213_5, epsilon = f32::from(f16::EPSILON));
+    }
+
     #[test]
     fn test_cosine_large() {
         let total = 1024;
         let x = (0..total).map(|v| v as f32).collect::<Vec<_>>();
         let y = (1024..1024 + total).map(|v| v as f32).collect::<Vec<_>>();
         let d = cosine_distance_batch(&x, &y, total).collect::<Vec<_>>();
-        assert_relative_eq!(d[0], cosine_dist_brute_force(&x, &y));
+        assert_relative_eq!(
+            d[0],
+            cosine_dist_brute_force(&x, &y),
+            epsilon = f32::from(f16::EPSILON) * 5.0
```

Comment:
If we changed the computation to aggregate into `f32`, could we reduce this epsilon to get more accuracy?

Reply:
Aggregating to f32 actually increases the error in the current test.